File: java/formula-java/src/main/java/org/formula/Resolvable.java
package org.formula;

import java.util.List;
import org.formula.context.DataContext;

public interface Resolvable {

    static Resolvable just(String value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(int value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(double value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(boolean value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(ResolvedValue value) {
        return StaticResolvable.of(value);
    }

    static Resolvable empty() {
        return EmptyResolvable.INSTANCE;
    }

    default ResolvedValue resolve() {
        return resolve(DataContext.EMPTY);
    }

    ResolvedValue resolve(DataContext context);

    String asFormula();

    static Resolvable concat(Resolvable... values) {
        return concat(List.of(values));
    }

    static Resolvable concat(Iterable<Resolvable> values) {
        return new ResolvableList(values);
    }
}

File: cs/formula-cs/Formula/Resolvable.cs
namespace Formula;

public interface IResolvable
{
    ResolvedValue Resolve()
    {
        return Resolve(DataContext.Empty);
    }
    
    ResolvedValue Resolve(IDataContext context);
}

public static class Resolvable
{
    public static IResolvable Empty { get; } = new EmptyResolvable();
    public static IResolvable Just(ResolvedValue value)
    {
        return StaticResolvable.Of(value);
    }
    
    public static IResolvable Just(string value)
    {
        return Just(ResolvedValue.Of(value));
    }
    
    public static IResolvable Just(int value)
    {
        return Just(ResolvedValue.Of(value));
    }
    
    public static IResolvable Just(double value)
    {
        return Just(ResolvedValue.Of(value));
    }
    
    public static IResolvable Just(bool value)
    {
        return Just(ResolvedValue.Of(value));
    }
}

public class StaticResolvable : IResolvable
{
    private readonly ResolvedValue _value;

    public static IResolvable Of(ResolvedValue value)
    {
        return new StaticResolvable(value);
    }

    public ResolvedValue Resolve()
    {
        return _value;
    }

    public ResolvedValue Resolve(IDataContext context)
    {
        return Resolve();
    }
    
    private StaticResolvable(ResolvedValue value)
    {
        _value = value;
    }
}

internal class EmptyResolvable : IResolvable
{
    public ResolvedValue Resolve(IDataContext context)
    {
        return ResolvedValue.None;
    }
}